velox/functions/prestosql/Zip.cpp

```diff
@@ -0,0 +1,221 @@
+      rawResultNulls[i] = resultNulls[i]->asMutable<uint64_t>();
+    }
+
+    // Final Row Vector
+    const auto finalArraySize = rows.end();
+    BufferPtr finalArrayIndices = allocateIndices(finalArraySize, pool);
+    BufferPtr finalArrayNulls =
+        AlignedBuffer::allocate<bool>(finalArraySize, pool, bits::kNotNull);
+    BufferPtr finalArrayOffsets = allocateOffsets(finalArraySize, pool);
+    BufferPtr finalArraySizes = allocateSizes(finalArraySize, pool);
+
+    auto rawFinalArrayOffsets = finalArrayOffsets->asMutable<vector_size_t>();
+    auto rawFinalArraySizes = finalArraySizes->asMutable<vector_size_t>();
+
+    // Create right offsets/indexes for the individual and final result arrays.
+    int rowIndex = 0;
+    int finalResultIndex = 0;
+    rows.applyToSelected([&](auto row) {
+      // Get the max size for that row.
+      auto maxRowSize = maxArraySizes[row];
+      rawFinalArrayOffsets[finalResultIndex] = rowIndex;
+
+      for (int arr = 0; arr < numInputArrays; arr++) {
+        auto arrayOffset = rawOffsets[arr][indices[arr][row]];
+        auto vectorSize = rawSizes[arr][indices[arr][row]];
+        for (int i = 0; i < maxRowSize; i++) {
+          if (i < vectorSize) {
+            rawResultIndices[arr][rowIndex + i] = arrayOffset++;
+          } else {
+            bits::setNull(rawResultNulls[arr], rowIndex + i, true);
+          }
+        }
+      }
+      rowIndex += maxRowSize;
+      rawFinalArraySizes[finalResultIndex] = maxRowSize;
+      finalResultIndex++;
+    });
+
+    // Now create result dictionary vectors.
+    std::vector<VectorPtr> resultDictionaryVectors(numInputArrays);
+
+    for (int i = 0; i < numInputArrays; i++) {
+      resultDictionaryVectors[i] = BaseVector::wrapInDictionary(
+          resultNulls[i],
+          resultIndices[i],
+          rowIndex,
+          baseVectors[i]->elements());
+    }
+
+    auto rowType = outputType->childAt(0);
+    auto rowVector = std::make_shared<RowVector>(
+        pool,
+        rowType,
+        BufferPtr(nullptr),
+        resultElementsSize,
+        resultDictionaryVectors);
+
+    // Now convert these to an Array
+    auto arrayVector = std::make_shared<ArrayVector>(
+        pool,
+        outputType,
+        finalArrayNulls,
+        finalResultIndex,
+        finalArrayOffsets,
+        finalArraySizes,
+        std::move(rowVector));
+
+    context->moveOrCopyResult(arrayVector, rows, result);
+  }
+
+  static std::vector<std::shared_ptr<exec::FunctionSignature>> signatures() {
+    std::vector<std::shared_ptr<exec::FunctionSignature>> signatures;
+    signatures.reserve(kMaxArity - kMinArity);
+
+    // Build all signatures from kMinArity to kMaxArity.
+    for (int i = kMinArity; i < kMaxArity; i++) {
+      auto builder = exec::FunctionSignatureBuilder();
+      std::vector<std::string> allTypeVars;
+      allTypeVars.reserve(i);
+
+      for (int j = 0; j < i; j++) {
+        auto typeVar = std::to_string('A' + j);
```

Comment:
nit: it might be clearer to create a vector of types names of size kMaxArity outside of the loop and re-use it.
- consider using fmt::format for consistency
- consider renaming A to E since the type name refers to the type of the array element

```
std::vector<std::string> elementTypeNames(kMaxArity);
for (auto i = 0; i < kMaxArity; ++i) {
   elementTypeNames[i] = fmt::format("E{}", i);
}

builder.typeVariable(elementTypeNames[j]);
....
```